### ISSUE TITLE
Make directories configurable

### ISSUE THREAD

**Issue by tuxpeople** [Kaiede/Bedrockifier#84]
**Is your feature request related to a problem? Please describe.**
I'd like to have one directory, where I can save all the Minecraft backups. Regardless of the server / container they are coming from. Right now, all backups are going to `/backups` which is also the place where `config.yml` is located and where Bedrockifier does create a `.service_is_healthy` file. When I point multiple Bedrockifier to the same backup folder,  I can dynamically mount the correct `config.yml` into the folder. But all of them are seeing the same `.service_is_healthy`. I assume, that's not the best case. 

This feature may also be interesting for people who'd like to have the `config.yaml` on their docker hosts, but the backup on another machine like a NAS.

**Describe the solution you'd like**
I'd like to be able to specify a folder (or subfolder relative to `/backups`) for the `*.mcworld` files in `config.yml`.

**Describe alternatives you've considered**
Currently, I've no other idea. I'm also not sure what happens if Bedrockifier finds `*.mcworld` files from other instances in that folder. Maybe that can cause issues as well?

**Additional context**
Maybe I miss something here, so this is my current setup: I'm running multiple Minecraft Bedrock servers in Kubernetes. I'm using the helm charts from itzg (https://github.com/itzg/minecraft-server-charts) to deploy them. Right now, I can configure the SSH stuff there, but I'm not able to access SSH from outside the pod (a pod is an Kubernetes object which runs one or multiple containers in it). But I've submitted a pull request for this: https://github.com/itzg/minecraft-server-charts/pull/191

Meanwhile, I add Betrockifier as a sidecar container to the Bedrock server pod. Therefore, in my `config.yml` the line for the SSH connection always looks like this: `ssh: 127.0.0.1:2222`. This would prevent me to add multiple containers to one `config.yml` (as well as the fact that I can't connect to another Bedrock instance because of the issue described above). So I end up with multiple `config.yml` but I'd like to have one folder for all the `*.mcworld` files.

**Comment by Kaiede** [comment]
Interesting suggestion, just a couple notes on your specific thoughts below for context:

> I'm also not sure what happens if Bedrockifier finds *.mcworld files from other instances in that folder. Maybe that can cause issues as well?

It will try to trim them, so there can definitely be a race condition which isn’t ideal if multiple backup services are being used with a single output directory, or worse if the trim settings are different on each (most restrictive wins). The service currently doesn’t expect that there may be output files that it doesn’t own in the backup directory. And in the situation where someone might use the same server, but refresh the world, knowing what worlds the service is allowed to trim is rather important. 

> Meanwhile, I add Betrockifier as a sidecar container to the Bedrock server pod. Therefore, in my config.yml the line for the SSH connection always looks like this: ssh: 127.0.0.1:2222.

Does kubernetes not offer private networks like docker compose does, where individual services would have a named address that could be used, even though the host itself is all the same address?

**Comment by Kaiede** [comment]
I will add there is rudimentary support for this using the backupsPath setting in the config file, it’s just not documented.

Trim would still be a concern but might unblock you in the short term?

**Comment by tuxpeople** [comment]
> Interesting suggestion, just a couple notes on your specific thoughts below for context:

Thanks. Believe me, I'm full of interesting suggestions :-) But I'm aware that this project's goal isn't to add all my special wishes ;)

> > I'm also not sure what happens if Bedrockifier finds *.mcworld files from other instances in that folder. Maybe that can cause issues as well?
> 
> It will try to trim them, so there can definitely be a race condition which isn’t ideal if multiple backup services are being used with a single output directory, or worse if the trim settings are different on each (most restrictive wins). The service currently doesn’t expect that there may be output files that it doesn’t own in the backup directory. And in the situation where someone might use the same server, but refresh the world, knowing what worlds the service is allowed to trim is rather important.

After thinking more carefully about it, I also found this challenge: The filenames are the world name, followed by the date/time. We cannot use the world name to filter which files we need to trim, as it might be possible to have the same world name on different servers. So we would have to change how the filename is being constructed, eg. prepending it with the name of the server.

> > Meanwhile, I add Betrockifier as a sidecar container to the Bedrock server pod. Therefore, in my config.yml the line for the SSH connection always looks like this: ssh: 127.0.0.1:2222.
> 
> Does Kubernetes not offer private networks like docker compose does, where individual services would have a named address that could be used, even though the host itself is all the same address?

The pod itself has a hostname, which I could use. But it's not fully predictable, as it contains a randomly generated suffix/postfix (?). I could use it if I could use environment variables in `config.yml`.

From outside, I have a Kubernetes object called "Service" which I can use to point to one (or multiple) pod(s). You can compare it to something like a reverse proxy. And here I have a predictable name. This is what I would use as soon as my PR gets merged into the helm chart.

> I will add there is rudimentary support for this using the backupsPath setting in the config file, it’s just not documented.
> 
> Trim would still be a concern but might unblock you in the short term?

It's not a blocker, as I still can use multiple directories to backup. So no need for a hotfix here, we can discuss the issue further to have a good solution. If you think this would be an good and useful feature.

**Comment by Kaiede** [comment]
> It's not a blocker, as I still can use multiple directories to backup. So no need for a hotfix here, we can discuss the issue further to have a good solution. If you think this would be an good and useful feature.

It wouldn't be a hotfix. The service is implemented such that there are more config variables (and environment variables) than are documented. These are all "advanced" settings these days that generally should never need be touched, but exist for either backwards compatibility from some of the really early stuff (from before there was a config file at all), or in the off chance it was needed. 

`backupPath` in the config.yml will change where the service puts the backup files, but it does need to be a path that exists.

**Comment by Kaiede** [comment]
So, I just want to follow up on this, because I think this is still (mostly) doable without changes from my end. 

Each Host
- backups/
  - config.yml
  - .authorizedKeys
  - remoteFolder/ (mapped?)
    - nodeName/
      - .service_is_healthy
      - <backups for this node>

In the configuration file, you would then need you to set `backupPath: "/backups/remoteFolder/nodeName/"`. With this, you'd be able to do offsite backups of the remote folder.

Another approach where you can use two mapped volumes will work as well. DATA_DIR controls where the service looks for things like config.yml and .authorizedKeys (new file). It also happens to set the backup directory, but with `backupPath` that can be pointed back to /backups into a subdirectory if you like.

Environment:
```
        DATA_DIR: "/config"
```

Volumes:
```
      # Map configuration folder into /config
      - <config volume>:/config
      # Map your backups folder into /backups
      - <your volume>:/backups
      # Map the data folders for both servers
      <snip>
```

config.yml:
```
backupPath: "/backups/nodeName"
```

This still leaves a couple things I can do on my end to make this cleaner in the long term though:

- Make it _easier_ to define a separate config directory, and detect it automatically if /config exists. 
- Either move .service_is_healthy to the config directory, or move it completely internal to the container as it is a temp file.

**Comment by Kaiede** [comment]
I've made changes that are now in the test tag to make this easier.

- /config is a new directory that if you map it to a folder with config.yml, it will also be used for service-related files (.authorizedKeys, .service_is_healthy). If you don't, it will fallback to use the same path as the backup files like it does today.
- /data or /backups can be mapped, and will be used for backup files.

**Comment by Kaiede** [comment]
I'll just add that I've been using this feature to push my storage of backups onto a NAS using an CIFS-based docker volume, while the config files are stored locally. This feedback has made my own setup simpler and nicer to use (no more duplicati on the server nodes themselves, less disk space needed on the nodes as well), so thanks.
